### PR TITLE
Add entire _dynamo.config as a json for logging

### PR DESCRIPTION
**Pull request by jovianjaison**
cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: torch/_dynamo/convert_frame.py [hunk 1]
```diff
@@ -1092,6 +1100,7 @@ def format_guard_failures() -> str:
                 config.suppress_errors,
                 config.inline_inbuilt_nn_modules,
                 config.specialize_float,
+                str(json.dumps(config_dict)),
```
Comment (Skylion007) [inline torch/_dynamo/convert_frame.py:1103]:
```suggestion
                json.dumps(config_dict),
```
Should already return str?

File: torch/_dynamo/convert_frame.py [hunk 2]
```diff
@@ -1059,6 +1060,13 @@ def format_guard_failures() -> str:
                 torch._logging.get_structured_logging_overhead()
             )
 
+            def convert_sets_to_strings(d: Dict[str, Any]) -> Dict[str, Any]:
+                return {
+                    key: str(value) if isinstance(value, set) else value
```
Comment (ezyang) [inline torch/_dynamo/convert_frame.py:1065]: Wouldn't a list be better here?

**Comment by ezyang** [comment]
How are you testing this?

File: torch/_dynamo/convert_frame.py [hunk 3]
```diff
@@ -1092,6 +1100,7 @@ def format_guard_failures() -> str:
                 config.suppress_errors,
                 config.inline_inbuilt_nn_modules,
                 config.specialize_float,
+                json.dumps(config_dict),
```
Comment (ezyang) [inline torch/_dynamo/convert_frame.py:1103]: I don't see evidence that any filtering has happened here; how big is this string right now?

**Comment by ezyang** [comment]
It would be really great for this dump to also make its way to tlparse via trace_structured

File: torch/_dynamo/convert_frame.py (hunk 3 above)
Comment (jovianjaison) [reply torch/_dynamo/convert_frame.py:1103]: Yes, updated to list.

File: torch/_dynamo/convert_frame.py (hunk 2 above)
Comment (jovianjaison) [reply torch/_dynamo/convert_frame.py]: Yes, updated to list.

**Comment by jovianjaison** [comment]
> How are you testing this?

$ python test/dynamo/test_structured_trace.py
ssss{"TYPE_CHECKING": false, "log_file_name": null, "verbose": false, "verify_correctness": false, "minimum_call_count": 1, "dead_code_elimination": true, "cache_size_limit": 8, "accumulated_cache_size_limit": 256, ...

**Comment by jovianjaison** [comment]
> It would be really great for this dump to also make its way to tlparse via trace_structured

Yeah I can try working on a PR for that.

**Comment by jovianjaison** [comment]
> How are you testing this?

Printed the json dump and ran
$ python test/dynamo/test_structured_trace.py
ssss{"TYPE_CHECKING": false, "log_file_name": null, "verbose": false, "verify_correctness": false, "minimum_call_count": 1, "dead_code_elimination": true, "cache_size_limit": 8, "accumulated_cache_size_limit": 256, ...

File: torch/_dynamo/convert_frame.py (hunk 2 above)
Comment (ezyang) [reply torch/_dynamo/convert_frame.py:1065]: This is exactly why you need filtering! Don't just put garbage data in Scuba, it's not free

**Comment by ezyang** [comment]
> Printed the json dump and ran

Can you post the actual full json dump?

**Comment by jovianjaison** [comment]
> > Printed the json dump and ran
> 
> Can you post the actual full json dump?

{"TYPE_CHECKING": false, "log_file_name": null, "verbose": false, "verify_correctness": false, "minimum_call_count": 1, "dead_code_elimination": true, "cache_size_limit": 8, "accumulated_cache_size_limit": 256, "skip_code_recursive_on_cache_limit_hit": true, "fail_on_cache_limit_hit": false, "specialize_int": false, "specialize_float": true, "dynamic_shapes": true, "use_lazy_graph_module": true, "assume_static_by_default": true, "automatic_dynamic_shapes": true, "force_parameter_static_shapes": true, "force_nn_module_property_static_shapes": true, "allow_ignore_mark_dynamic": false, "guard_nn_modules": true, "guard_nn_modules_using_dict_tags": true, "traceable_tensor_subclasses": "set()", "replay_record_enabled": false, "rewrite_assert_with_torch_assert": true, "disable": false, "cprofile": false, "skipfiles_inline_module_allowlist": {}, "allowed_functions_module_string_ignorelist": "{'torch._refs', 'torch.testing', 'torch._decomp', 'torch._prims', 'torch.distributions'}", "repro_after": null, "repro_level": 2, "repro_forward_only": false, "repro_tolerance": 0.001, "repro_ignore_non_fp": false, "same_two_models_use_fp64": true, "capture_scalar_outputs": false, "capture_dynamic_output_shape_ops": false, "prefer_deferred_runtime_asserts_over_guards": false, "allow_complex_guards_as_runtime_asserts": false, "force_unspec_int_unbacked_size_like_on_torchrec_kjt": false, "enforce_cond_guards_match": true, "optimize_ddp": true, "do_not_emit_runtime_asserts": false, "_ddp_optimization_mode": ["ddp_optimizer", "python_reducer", "python_reducer_without_compiled_forward", "no_optimization"], "skip_torchrec": true, "optimize_ddp_lazy_compile": false, "skip_fsdp_guards": true, "skip_fsdp_hooks": true, "skip_nnmodule_hook_guards": true, "raise_on_ctx_manager_usage": true, "raise_on_unsafe_aot_autograd": false, "error_on_nested_jit_trace": true, "error_on_nested_fx_trace": true, "allow_rnn": false, "capture_sparse_compute": true, "error_on_recompile": false, "report_guard_failures": true, "base_dir": "/home/jovian/jovian_github/pytorch", "trace_numpy": true, "numpy_default_float": "float64", "numpy_default_complex": "complex128", "numpy_default_int": "int64", "use_numpy_random_stream": false, "enable_cpp_guard_manager": true, "inline_inbuilt_nn_modules": true, "record_compile_time_instruction_count": false, "debug_dir_root": "/home/jovian/jovian_github/pytorch/torch_compile_debug", "_save_config_ignore": "{'repro_after', 'skipfiles_inline_module_allowlist', 'repro_level', 'constant_functions'}", "cudagraph_backend_keep_input_mutation": false, "cudagraph_backend_support_input_mutation": false, "only_allow_pt2_compliant_ops": false, "capture_autograd_function": true, "capture_func_transforms": true, "log_compilation_metrics": false, "reorderable_logging_functions": "set()", "inject_BUILD_SET_unimplemented_TESTING_ONLY": false, "_autograd_backward_strict_mode_banned_ops": ["stride", "requires_grad", "storage_offset", "layout", "data", "is_coalesced", "is_complex", "is_conj", "is_contiguous", "is_cpu", "is_cuda", "is_distributed", "is_floating_point", "is_inference", "is_ipu", "is_leaf", "is_maia", "is_meta", "is_mkldnn", "is_mps", "is_mtia", "is_neg", "is_nested", "is_nonzero", "is_pinned", "is_quantized", "is_same_size", "is_set_to", "is_shared", "is_signed", "is_sparse", "is_sparse_csr", "is_vulkan", "is_xla", "is_xpu"], "fake_tensor_cache_enabled": true, "fake_tensor_cache_crosscheck_enabled": false, "compiled_autograd": false, "enable_compiler_collectives": false, "suppress_errors": false}

File: torch/_dynamo/convert_frame.py (hunk 2 above)
Comment (jovianjaison) [reply torch/_dynamo/convert_frame.py]: Makes sense, added filtering to exclude sets  containing _TensorMeta and functions.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
By the way, looking over your dump, these would also be good to delete: "TYPE_CHECKING", "log_file_name", "verbose", "repro_after", "repro_level", "repro_forward_only", "repro_tolerance", "repro_ignore_non_fp", "same_two_models_use_fp64", "base_dir", "debug_dir_root", "_save_config_ignore", "log_compilation_metrics", "inject_BUILD_SET_unimplemented_TESTING_ONLY", "_autograd_backward_strict_mode_banned_ops"

But now I'm kind of wondering if we should control this via sitevar. Well, I don't think we're going to blow up the log entry size if we log these, so we can do it for now.

**Comment by ezyang** [comment]
@masnesral may also be of interest to you

**Pull request closed without merging**